src/datascience-ui/data-explorer/mainPanel.tsx

```diff
@@ -304,17 +309,60 @@ export class MainPanel extends React.Component<IMainPanelProps, IMainPanelState>
         return [];
     }
 
-    private normalizeRows(rows: JSONArray): ISlickRow[] {
+    private normalizeData(rows: JSONArray): ISlickRow[] {
+        // While processing rows we may encounter Inf, -Inf or NaN.
+        // These rows' column types will initially be 'string' or 'object' so
+        // make sure we update the column types
+        // Set of columns to update based on this batch of rows
+        const columnsToUpdate = new Set<string>();
         // Make sure we have an index field and all rows have an item
-        return rows.map((r: any | undefined) => {
+        const normalizedRows = rows.map((r: any | undefined) => {
             if (!r) {
                 r = {};
             }
             if (!r.hasOwnProperty(this.state.indexColumn)) {
                 r[this.state.indexColumn] = uuid();
             }
+            for (let [key, value] of Object.entries(r)) {
+                switch (value) {
+                    case 'nan':
+                        r[key] = NaN;
+                        if (!this.columnsContainingInfOrNaN.has(key)) {
+                            columnsToUpdate.add(key);
+                            this.columnsContainingInfOrNaN.add(key);
+                        }
+                        break;
+                    case '-inf':
+                        r[key] = -Infinity;
+                        if (!this.columnsContainingInfOrNaN.has(key)) {
+                            columnsToUpdate.add(key);
+                            this.columnsContainingInfOrNaN.add(key);
+                        }
+                        break;
+                    case 'inf':
+                        r[key] = Infinity;
+                        if (!this.columnsContainingInfOrNaN.has(key)) {
+                            columnsToUpdate.add(key);
+                            this.columnsContainingInfOrNaN.add(key);
+                        }
+                        break;
+                    default:
+                }
+            }
             return r;
         });
+        // Need to update the column types so that that column gets number treatment.
+        // This should be unusual in practice.
```

Comment:
Considering if this is actually something that would happen or not, but in a rare case could this be converting something like a string column to a number column? If it happened to have a string value of 'nan' or 'inf' in it? Not sure if that possibility is worth guarding against. You know the data best here Joyce, possible issue? Or not something worth worrying about.

Reply:
It's possible because data can be heterogenous. Our code today assumes ColumnType applies to all the data in the column. I don't really know what the best way to handle this is. I originally replaced `np.inf` with `_VSCODE_inf` and so on in our Python scripts for that exact reason, and that worked in some cases (e.g. ndarray, tensor) but not others (e.g. DataFrame).

One option is to enable this check only for tensors, since it's especially critical to detect inf/-inf/nan gradients. In other scenarios it doesn't matter as much. But I know jmew wants this fixed globally so that wouldn't work.